### DISCUSSION TITLE
cursor.copy() results in "ProgrammingError - cannot adapt type 'int32' using placeholder '%t' (format: TEXT)"

### DISCUSSION THREAD

**Discussion by Raichuu41** [psycopg/psycopg#193] [Ideas]
When trying to push data to a table with the cursor.copy() write_row() method, I am getting following error: 
`psycopg.errors.QueryCanceled: COPY from stdin failed: error from Python: ProgrammingError - cannot adapt type 'int32' using placeholder '%t' (format: TEXT)`

The data is from a Pandas DataFrame which is extracted into records (list of tuples) and the columns (as a comma-separated string). 
Assume a database table called "mytable" in the schema "myschema". The table only has the column "all_ids" which is of data type integer. Assume a Pandas DataFrame with the column "all_ids" and the values "1, 2, 3, 4, 5"

DataFrame sample data:
```python
example = [1, 2, 3, 4, 5]
df = pandas.DataFrame({"all_ids": pandas.Series(example, dtype="int32")})
```

Attempt to save it to the database:
```python
records = [tuple(x) for x in df.to_numpy()]
cursor = connection.cursor()  # assume connection variable is already there
with cursor.copy(f'COPY myschema.mytable (all_ids) FROM STDIN) as copy:
    for record in records:
        copy.write_row(record)
```

This leads to the error but I don't understand what's the issue? Interestingly, when changing the data values to be of type str `pandas.Series(example, dtype="str")`, the error no longer happens and the column is still stored as integer in the database table. Is there any kind of limitation that I am not aware of yet for the PostgrreSQL COPY operation or is this a bug in psycopg?

**Comment by dvarrazzo** [comment]
Hello,

it is simply that psycopg doesn't currently support numpy data types. 

```python
>>> i = numpy.int32(123)
>>> cnn.execute("select %s", [i])
...
ProgrammingError: cannot adapt type 'int32' using placeholder '%s' (format: AUTO)
```

They can be added. As a quick test, reusing the same dumper used for Python works ok:

```python
>>> cnn.adapters.register_dumper(type(i), psycopg.types.numeric.IntDumper)
>>> cur = cnn.execute("select %s", [i])
>>> cur.fetchone()
```

See [the docs](https://www.psycopg.org/psycopg3/docs/advanced/adapt.html) for info about types customization.

Proper support to the whole mypy data model is probably a bigger work, however it is surely something useful to have, and the result could be pretty efficient (especially using binary dumpers/loaders).

**Comment by dvarrazzo** [comment]
The reason why using `dtype="str"` works is probably because the resulting Python object is returned as a normal string. This is passed to Postgres as an untyped literal, and Postgres cast rules convert it into the target type.

The `numpy.int32`, instead, is not a subclass of the Python `int`, so Psycopg has no clue about how to convert it.

```python
>>> numpy.int32.__mro__
(numpy.int32,
 numpy.signedinteger,
 numpy.integer,
 numpy.number,
 numpy.generic,
 object)
```

**Comment by Raichuu41** [comment]
Thank you very much for the detailed and quick response! Now I have two more questions: What is the best approach to add custom support for the int32/64 (and probably the float ones) numpy data types? If registering the numeric dumpers from psycopg for the specific data types is already enough, is that the better approach than type casting the specific column in the DataFrame to str?

Also one more question that arises due to it: If including the same data types (int32 or int64) in a DataFrame but with more columns than only the specific one, the same code to push the data doesn't raise an error and it works as expected, as if there was numpy datatypes support. What may be the reason for this behaviour?
To give a specific sample in one of my tables: It includes the data types "varchar, integer, date" (in total 14 columns), the data is loaded into a DataFrame with the same approach but without the enforced type casting
```python
my_data = [{'col1': 'abc', 'col2': 123, 'col3': '2021-03-19', ...}, 
           {......}]
df = pandas.DataFrame(my_data)  # col2 gets the data type int64
``` 
And pushing this DataFrame with the same copy code from my first post, it works without errors.

**Comment by dvarrazzo** [comment]
In order to answer these questions, it's best keeping in mind what is the fundamental mechanism behind dumping python to postgres. Psycopg first looks at the type of the object dumped. If a dumper is registered for that type, it uses it. If a dumper is registered for a subclass, it uses it (from the most specific to the most generic). 

For certain types, looking at the type is not enough, for instance a Python int can be a postgres int2, 4, 8 or numeric. In this case the dumper has a more advanced mechanism of specialization.

So, in order to answer your questions, we need to know the types of the object in question.

> What is the best approach to add custom support for the int32/64 (and probably the float ones) numpy data types? If registering the numeric dumpers from psycopg for the specific data types is already enough, is that the better approach than type casting the specific column in the DataFrame to str?

Registering the dumpers is a good approach, and what we did earlier works because the textual representation of the numbers is always the same.

In case of limited range ints it would be probably more efficient to use a more specific dumper: for the text mode, `psycopg.types.numeric.Int4Dumper` is a perfect fit for `numpy.int32` because they are both 4 bytes signed int, using it will do less work than the `IntDumper` proposed earlier. Using specific types is probably better than using strings.

> the same code to push the data doesn't raise an error and it works as expected, as if there was numpy datatypes support. What may be the reason for this behaviour?

What are the types of the Python objects you end up writing? That will answer. They might be stdlib python types, already supported by psycopg, or their subclasses.

**Reply by Raichuu41** [reply]
The other data types are as following, taken from data_frame.dtypes output:
```
name                  object
ehraid                 int64
uid                   object
chid                  object
legal_seat_id          int64
legal_seat            object
register_office_id     int64
legal_form_id          int64
status                object
rab_id                object
shab_date             object
delete_date           object
purpose               object
timestamp             object
```

Pushing this dataframe to the database with the given code which uses the COPY operation, it works without errors even though there's the int64 data type included in 4 columns. But trying to push the dataframe sample given on the first post (with only one column of type int64), the push fails with the error message.

**Reply by dvarrazzo** [reply]
> The other data types are as following, taken from data_frame.dtypes output:

These are data types as defined for pandas/numpy: it would be good to know what Python types are. e.g. looking at `print([type(i) for i in record])`, for one of the records returned by `df.to_numpy()`.

**Reply by Raichuu41** [reply]
That's actually quite interesting, the classes are changing on that.
Here's the given output:
```python
[<class 'str'>, <class 'int'>, <class 'str'>, <class 'str'>, <class 'int'>, <class 'str'>, <class 'int'>, <class 'int'>, <class 'str'>, <class 'float'>, <class 'str'>, <class 'NoneType'>, <class 'str'>, <class 'datetime.date'>]
```

As we can see, the int64 columns are now only of class int. The timestamp (object) column is now of datetime.date as expected. So when converting the data frame with to_numpy() if there is more than one column, the int64 entries become type int?
That's weird behaviour from Pandas' to_numpy() function.
Taking the intial trouble example again, where there is only one column of data type int64, this is the output in each record:
```python
[<class 'numpy.int64'>]
```

**Comment by dvarrazzo** [comment]
Converted this into a discussion. Added #192 to track adding numpy types support to Psycopg.
